compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -7362,7 +7378,13 @@ private STNode parseLockStatement() {
         STNode lockKeyword = parseLockKeyword();
         STNode blockStatement = parseBlockNode();
         endContext();
-        return STNodeFactory.createLockStatementNode(lockKeyword, blockStatement);
+        STNode onFailClause;
+        if (peek().kind == SyntaxKind.ON_KEYWORD) {
+            onFailClause = parseOnFailClause();
+        } else {
+            onFailClause = STNodeFactory.createEmptyNode();
+        }
```

Comment:
We need to for error recovery here, if the next token is not the `on` keyword nor any statement-starting token.
eg: missing `on` keyword, but the rest is present from the `fail` keyword.

Reply:
Seems on on fail clause is not optional.